c/src/ml-api-service-query-client.c

```diff
@@ -0,0 +1,212 @@
+        data_s->tensors[i].size);
+  }
+
+  g_async_queue_push (mls->out_data_queue, copied_data);
+}
+
+/**
+ * @brief Creates query client service handle with given ml-option handle.
+ */
+int
+ml_service_query_create (ml_option_h option, ml_service_h * h)
+{
+  int status = ML_ERROR_NONE;
+
+  gchar *description = NULL;
+
+  ml_option_s *_option;
+  GHashTableIter iter;
+  gchar *key;
+  ml_option_value_s *_option_value;
+
+  GString *tensor_query_client_prop;
+  gchar *prop = NULL;
+
+  ml_service_s *mls;
+
+  _ml_service_query_s *query_s;
+  ml_pipeline_h pipe_h;
+  ml_pipeline_src_h src_h;
+  ml_pipeline_sink_h sink_h;
+  gchar *caps = NULL;
+  guint timeout = 1000U;        /* default 1s timeout */
+
+  check_feature_state (ML_FEATURE_SERVICE);
+  check_feature_state (ML_FEATURE_INFERENCE);
+
+  if (!option) {
+    _ml_error_report_return (ML_ERROR_INVALID_PARAMETER,
+        "The parameter, 'option' is NULL. It should be a valid ml_option_h, which should be created by ml_option_create().");
+  }
+
+  if (!h) {
+    _ml_error_report_return (ML_ERROR_INVALID_PARAMETER,
+        "The parameter, 'h' (ml_service_h), is NULL. It should be a valid ml_service_h.");
+  }
+
+  _option = (ml_option_s *) option;
+  g_hash_table_iter_init (&iter, _option->option_table);
+  tensor_query_client_prop = g_string_new (NULL);
+  while (g_hash_table_iter_next (&iter, (gpointer *) & key,
+          (gpointer *) & _option_value)) {
+    if (0 == g_ascii_strcasecmp (key, "host")) {
+      g_string_append_printf (tensor_query_client_prop, " host=%s ",
+          (gchar *) _option_value->value);
+    } else if (0 == g_ascii_strcasecmp (key, "port")) {
+      g_string_append_printf (tensor_query_client_prop, " port=%u ",
+          *((guint *) _option_value->value));
+    } else if (0 == g_ascii_strcasecmp (key, "dest-host")) {
+      g_string_append_printf (tensor_query_client_prop, " dest-host=%s ",
+          (gchar *) _option_value->value);
+    } else if (0 == g_ascii_strcasecmp (key, "dest-port")) {
+      g_string_append_printf (tensor_query_client_prop, " dest-port=%u ",
+          *((guint *) _option_value->value));
+    } else if (0 == g_ascii_strcasecmp (key, "connect-type")) {
+      g_string_append_printf (tensor_query_client_prop, " connect-type=%s ",
+          (gchar *) _option_value->value);
+    } else if (0 == g_ascii_strcasecmp (key, "topic")) {
+      g_string_append_printf (tensor_query_client_prop, " topic=%s ",
+          (gchar *) _option_value->value);
+    } else if (0 == g_ascii_strcasecmp (key, "timeout")) {
+      g_string_append_printf (tensor_query_client_prop, " timeout=%u ",
+          *((guint *) _option_value->value));
+      timeout = *((guint *) _option_value->value);
+    } else if (0 == g_ascii_strcasecmp (key, "caps")) {
+      caps = g_strdup (_option_value->value);
+    } else {
+      _ml_logw ("Ignore unknown key for ml_option: %s", key);
+    }
+  }
+
```

Comment:
Next PR: plz add protection for mandatory option str.

Reply:
Thanks, I added a commit to fix this. PTAL.